File: src/leetcode/array/AddTwoNumbers.java
package leetcode.array;

public class AddTwoNumbers {

    public static ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode listNode = new ListNode(0);
        int preDiv = 0;

        ListNode curNode = listNode;

        while (l1 != null && l2 != null) {
            if(l1.next != null || l2.next!=null){
                curNode.next = new ListNode(0);
            }
            int curSum = l1.val + l2.val + preDiv;
            preDiv = curSum / 10;
            curNode.val = curSum % 10;
            curNode = curNode.next;
            l1 = l1.next;
            l2 = l2.next;
        }

        if (l1 == null) {
            while (l2 != null) {
                if(l2.next != null){
                    curNode.next = new ListNode(0);
                }
                int curSum = l2.val + preDiv;
                preDiv = curSum / 10;
                curNode.val = curSum % 10;
                curNode = curNode.next;
                l2 = l2.next;
            }
        } else {
            while (l1 != null) {
                if(l1.next !=null){
                    curNode.next = new ListNode(0);
                }
                int curSum = l1.val + preDiv;
                preDiv = curSum / 10;
                curNode.val = curSum % 10;
                curNode = curNode.next;
                l1 = l1.next;
            }
        }

        return listNode;
    }

    public static void main(String[] args) {
        ListNode l1 = new ListNode(2);
        ListNode l2 = new ListNode(4);
        ListNode l3 = new ListNode(3);
        ListNode l4 = new ListNode(9);
        l1.next = l2;
        l2.next = l3;
        l3.next = l4;

        ListNode a1 = new ListNode(5);
        ListNode a2 = new ListNode(6);
        ListNode a3 = new ListNode(4);
        a1.next = a2;
        a2.next = a3;

        System.out.println(AddTwoNumbers.addTwoNumbers(l1, a1));



    }
}

File: scala/two-sum/src/main/scala/AddTwoNumbers.scala
/**
  * https://leetcode.com/problems/add-two-numbers/
  * You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.
  *
  * You may assume the two numbers do not contain any leading zero, except the number 0 itself.
  *
  * Example:
  *
  * Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
  * Output: 7 -> 0 -> 8
  * Explanation: 342 + 465 = 807.
  */
object AddTwoNumbers {

  /*
   * Use carried to represent if there is a carry at current position.
   * Store the result just in the first node.
   */
  def addTwoNumbers(l1: ListNode, l2: ListNode): ListNode = {
    var carry = 0
    var p = new ListNode()
    var q = new ListNode()
    p.next = l1
    q.next = l2

    // Two numbers
    while (p.next != null && q.next != null) {
      p = p.next
      q = q.next
      var cur = carry + p.x + q.x
      carry = cur / 10
      p.x = cur % 10
    }

    // Use first node
    if (q.next != null)
      p.next = q.next

    // One number
    while (p.next != null) {
      p = p.next
      var cur = carry + p.x
      carry = cur / 10
      p.x = cur % 10
    }

    // The highest position
    if (carry == 1)
      p.next = new ListNode(1)

    l1
  }
}

class ListNode(var _x: Int = 0) {
  var next: ListNode = null
  var x: Int = _x
}
